cli/tests/test_op_revert_command.rs

```diff
@@ -0,0 +1,639 @@
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    main: qpvuntsm 1e742089 (empty) CC
+      @origin (ahead by 1 commits, behind by 1 commits): qpvuntsm hidden d9a9f6a0 (empty) BB
+    [EOF]
+    ");
+}
+
+#[test]
+fn test_bookmark_track_untrack_revert() {
+    let test_env = TestEnvironment::default();
+    test_env.add_config(r#"revset-aliases."immutable_heads()" = "none()""#);
+    let git_repo_path = test_env.env_root().join("git-repo");
+    git::init_bare(git_repo_path);
+    test_env
+        .run_jj_in(".", ["git", "clone", "git-repo", "repo"])
+        .success();
+    let work_dir = test_env.work_dir("repo");
+
+    work_dir.run_jj(["describe", "-mcommit"]).success();
+    work_dir
+        .run_jj(["bookmark", "create", "-r@", "feature1", "feature2"])
+        .success();
+    work_dir.run_jj(["git", "push", "--allow-new"]).success();
+    work_dir
+        .run_jj(["bookmark", "delete", "feature2"])
+        .success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+      @origin: qpvuntsm bab5b5ef (empty) commit
+    feature2 (deleted)
+      @origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+
+    // Track/untrack can be reverted so long as states can be trivially merged.
+    work_dir
+        .run_jj(["bookmark", "untrack", "feature1@origin", "feature2@origin"])
+        .success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+    feature1@origin: qpvuntsm bab5b5ef (empty) commit
+    feature2@origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+
+    work_dir.run_jj(["op", "revert"]).success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+      @origin: qpvuntsm bab5b5ef (empty) commit
+    feature2 (deleted)
+      @origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+
+    work_dir.run_jj(["op", "revert"]).success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+    feature1@origin: qpvuntsm bab5b5ef (empty) commit
+    feature2@origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+
+    work_dir
+        .run_jj(["bookmark", "track", "feature1@origin"])
+        .success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+      @origin: qpvuntsm bab5b5ef (empty) commit
+    feature2@origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+
+    work_dir.run_jj(["op", "revert"]).success();
+    insta::assert_snapshot!(get_bookmark_output(&work_dir), @r"
+    feature1: qpvuntsm bab5b5ef (empty) commit
+    feature1@origin: qpvuntsm bab5b5ef (empty) commit
+    feature2@origin: qpvuntsm bab5b5ef (empty) commit
+    [EOF]
+    ");
+}
+
+#[test]
+fn test_revert_latest_revert_implicitly() {
```

Comment:
This should be left in test_undo.rs because it tests the warning message of double undo.